Re-theme the default layout and home page

amitaibu opened the pull request:
This is still WIP and PR Depends on #133, but I open it to see if there is enough interest, so I could invest more time.

The motivation for this PR is to improve the first impression devs get when scaffolding for the first time, and on the same time provide some of the typical functionality of a website (along with my non-hidden agenda is also to improve my Yesod skills by getting reviews 😉 )

Here's the [diff](https://github.com/amitaibu/yesod-scaffold/compare/auth-dummy...amitaibu:semantic-ui) until #133 is merged

![joxi_screenshot_1473197789630](https://cloud.githubusercontent.com/assets/125707/18291885/28c40028-7493-11e6-8345-ecb086073b54.png)

![welcome_to_yesod_](https://cloud.githubusercontent.com/assets/125707/18291794/cd2de2a6-7492-11e6-8d63-796630e2455f.jpg)

amitaibu commented on Foundation.hs:
```diff
@@ -23,6 +25,16 @@ data App = App
     , appLogger      :: Logger
     }
 
+data MenuItem = MenuItem
```
what would be the best place to move those types?

amitaibu replied:
Should this stay in `Foundation.hs` or is there a better place for it?

snoyberg commented:
Probably worth emailing the Yesod mailing list about it. From my experience, Bootstrap is still very popular, and likely what most people would want in the scaffold.

amitaibu commented:
Sure, I can keep bootstrap, but still modernize the look + add the
functionality

On Sep 7, 2016 7:13 AM, "Michael Snoyman" <notifications@github.com> wrote:

> Probably worth emailing the Yesod mailing list about it. From my
> experience, Bootstrap is still very popular, and likely what most people
> would want in the scaffold.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/yesodweb/yesod-scaffold/pull/135#issuecomment-245170045>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAHrC_ACGqbin0JF9UhZzhUIjwcbz2DMks5qnjnWgaJpZM4J2UAa>
> .
>

amitaibu commented:
I _tried_  to do it with Bootstrap. I mean, I've been using Bootstrap for years, but my problem with the different MIT licenses [themes](https://bootswatch.com/), that even-though I'm able to easily replace the css file and get a new look and feel, in my eyes it all kinda looks the same.

![welcome_to_yesod_](https://cloud.githubusercontent.com/assets/125707/18326941/b30f2bdc-7550-11e6-93cd-937e00f14e37.jpg)

What I like about [SUI](http://semantic-ui.com/) (semantic UI) is that:

1. It gives a fresh look (= admitedly by the fact it's not as popular as bootstrap), not the bootstrapy one we're so used to. This IMO is inline with showing yesod as a modern framework
1. It's a matter of taste of course, but I think that for prototyping SUI allows having nicer UIs

Anyway, I'll sleep on it. Lets see how I feel about it in the morning 😉

amitaibu commented:
Ok, decided to stay with Bootstrap, so it would be easier to get it in 😄 

![joxi_screenshot_1473444993635](https://cloud.githubusercontent.com/assets/125707/18397891/eb0df71a-76d2-11e6-9671-627b844839f5.png)

MaxGabriel commented on templates/homepage.hamlet:
```diff
@@ -1,80 +1,141 @@
+            <h1 class="header">
+                Yesod; A modern framework for blazing fast websites
+            <h2>
+                Fast, Stable & spiced with great community
```
It's weird how Fast/Stable are capitalized but "spiced with a great community" is not

amitaibu replied:
Pushed a fix

MaxGabriel commented on templates/homepage.hamlet:
```diff
@@ -1,80 +1,141 @@
+    <div class="container">
+        <div class="row">
+            <h1 class="header">
+                Yesod; A modern framework for blazing fast websites
```
Maybe just me, but "Yesod; A" feels a little weird to me. I guess because "Yesod" isn't a full sentence/capitalizing A after a semicolon isn't right

amitaibu replied:
you mean you'd expect to have a lower case of the `a`?

MaxGabriel replied:
I'm not really sure what I'm expecting. Maybe a colon instead of a semicolon?

amitaibu replied:
The semicolon just looked kind of cool to me, but this can obviously change. I'm not a native English speaker, so I ben the English rules to my will 😜

MaxGabriel commented:
This redesign looks really good, nice work @amitaibu.

amitaibu commented:
Thanks @MaxGabriel  :)

There's one thing I know I wasn't consistent with - showing the classes in the divs - `<div .container>` vs `<div class="container">`. I actually prefer the latter, since it allows me to copy/ paste more easily.

We should probably be consistent, so I wonder if you folks agree with my comment above.

psibi commented on templates/homepage.julius:
```diff
@@ -22,6 +22,7 @@ $(function() {
         var newNode = $("<li></li>");
         newNode.text(data.message);
+        console.log(data);
         $("##{rawJS commentListId}").append(newNode);
```
@amitaibu You may want to remove the console part ?

amitaibu replied:
this is actually a feature not a bug ;) I have added this [text](https://github.com/yesodweb/yesod-scaffold/pull/135/files#diff-a8900c26af8ddefc60773f874e4506c4R124), so devs can see the response object.

psibi commented:
Nice work @amitaibu . This looks very good. I do feel that the semicolon looks a bit odd as Maximilian told. Also, I think `<div .container>` is preferable as it demonstrates Shakespeare capabilities. IIRC, the old code followed that format, but I don't have strong opinions.

amitaibu commented:
> Also, I think <div .container> is preferable as it demonstrates Shakespeare capabilities

It does, however IMO in terms of easily copy/ paste from other resources it's making it harder. But if that's what holds the PR I'll change it :)

chreekat commented:
On Mon, Sep 12, 2016 at 12:03:50PM -0700, Amitai Burstein wrote:
> > Also, I think <div .container> is preferable as it demonstrates
> > Shakespeare capabilities
> 
> It does, however IMO in terms of easily copy/ paste from other
> resources it's making it harder. But if that's what holds the PR
> I'll change it :)

I've come to think that Hamlet should not be considered as something
that is copy/pastable from other resources. :) Not even <pre> blocks
can be copy/pasted.

So why not

    <.container>

which also expands to <div class="container">.

amitaibu commented:
* I have replaced the `;` with an em-dash
* Converted HTML to use Shakespeare

![joxi_screenshot_1473876778265](https://cloud.githubusercontent.com/assets/125707/18524196/148dc4a8-7ac0-11e6-8c6e-42d67e415ad5.png)

Not sure why tests are failing, not sure it's related to this PR

psibi commented on test/Handler/HomeSpec.hs:
```diff
@@ -9,7 +9,7 @@ spec = withApp $ do
           get HomeR
           statusIs 200
-          htmlAllContain "h1" "Welcome to Yesod"
+          htmlAnyContain "h1" "Yesod; A modern framework for blazing fast websites"
```
I think the test will fail here because it still contains `;` but you have changed the content.

amitaibu replied:
Correct, I have fixed it. The tests still fail, but on what seems to look like an unrelated issue.

amitaibu commented:
Update: I see there's indeed one test that fails that I'll fix, but in [one environment](https://travis-ci.org/yesodweb/yesod-scaffold/jobs/160123518) the test even doesn't start

amitaibu commented:
`stack test` is passing
![build__260_-_yesodweb_yesod-scaffold_-_travis_ci](https://cloud.githubusercontent.com/assets/125707/18562240/929e0ae4-7b8b-11e6-993d-dd20a4188ec9.jpg)

 but the other one fails due to wrong dependency

![job__260_1_-_yesodweb_yesod-scaffold_-_travis_ci](https://cloud.githubusercontent.com/assets/125707/18562264/aa423e5e-7b8b-11e6-94e5-15386a48746a.jpg)

amitaibu commented:
@psibi do you happen to know how I can solve this? ^^

psibi commented:
The error seems to be happening because the latest lts uses
ghc-8.0.1. And we have restricted the use of base-4.9.0.0 in our cabal
file because of this bug: https://ghc.haskell.org/trac/ghc/ticket/12130

So, we have to wait till ghc 8.0.2 is released. Probably, till that we
can include that build in the allow_failures section with a proper
comment ?

snoyberg commented:
This is correct. I can update the branches now, and then merging in the changes should unblock you @amitaibu.

snoyberg commented:
OK, changes pushed, Travis will hopefully accept them.

amitaibu commented:
Thanks @snoyberg , re-rolled and tests have passed. So I guess it's ready for review 😄

psibi commented:
LGTM. CC: @MaxGabriel and others

psibi commented:
Thanks!

amitaibu commented:
Awesome!
